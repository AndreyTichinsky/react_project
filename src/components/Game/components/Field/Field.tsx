import React, { FC } from "react";
import styled from "@emotion/styled";
import type { FieldProps } from "types/field";

import { Cell } from "./components";

interface FieldWrapperStyle {
  cellSize: number;
  field: boolean[][];
}

const FieldWrapper = styled.div<FieldWrapperStyle>`
  position: relative;
  width: ${(props: FieldWrapperStyle) =>
    props.cellSize * props.field[0].length}px;
  height: ${(props: FieldWrapperStyle) =>
    props.cellSize * props.field.length}px;
  border: 1px solid #000;
`;

export const Field: FC<FieldProps> = ({ field, onClick, cellSize }) => (
  <FieldWrapper cellSize={cellSize} field={field}>
    {field.map((row, y) => [
      ...row.map((isAlive: boolean, x) => (
        <Cell
          key={`${y}_${x}`}
          x={x}
          y={y}
          cellSize={cellSize}
          isAlive={isAlive}
          onClick={onClick}
        />
      )),
    ])}
  </FieldWrapper>
);
