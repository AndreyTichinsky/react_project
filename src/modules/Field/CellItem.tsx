import styled from "@emotion/styled";
import { css } from "@emotion/core";

const BaseCell = css`
  border: 1px solid #000;
  box-sizing: border-box;
  width: 10px;
  height: 10px;
`;

const aliveCell = css`
  background-color: #000000;
`;

const deathCell = css`
  background-color: #ffffff;
`;

interface Props {
  isAlive: boolean;
  x: number;
  y: number;
  animationSpeed: number;
}

export const CellItem = styled.div`
  ${BaseCell};
  transition: background-color
    ${({ animationSpeed }: Props) => animationSpeed}ms ease;
  ${({ isAlive }: Props) => {
    return isAlive ? aliveCell : deathCell;
  }}
`;
