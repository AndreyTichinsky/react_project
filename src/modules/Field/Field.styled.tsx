import styled from "@emotion/styled";

interface FieldWrapperStyle {
  cellSize: number;
  field: boolean[][];
}

export const FieldWrapper = styled.div<FieldWrapperStyle>`
  position: relative;
  width: ${(props: FieldWrapperStyle) =>
    props.cellSize * props.field[0].length}px;
  height: ${(props: FieldWrapperStyle) =>
    props.cellSize * props.field.length}px;
  border: 1px solid #000;
`;
