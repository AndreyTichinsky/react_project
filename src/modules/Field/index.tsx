export { Field } from "./Field";
export { Cell } from "./Cell";
