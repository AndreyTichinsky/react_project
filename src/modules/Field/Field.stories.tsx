import React from "react";
import { Provider } from "react-redux";
import { store } from "@/redux/store";
import { withKnobs } from "@storybook/addon-knobs";
import { Field } from "./Field";

export default {
  title: "Field",
  decorators: [withKnobs],
};

export const emptyField = () => {
  return (
    <Provider store={store}>
      <Field />
    </Provider>
  );
};
