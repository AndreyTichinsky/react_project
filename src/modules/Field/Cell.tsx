import React, { FC, useCallback } from "react";
import { CellItem } from "./CellItem";
import { connect } from "react-redux";
import { AppDispatch } from "@/redux/store";
import { State } from "@/redux/reducers";
import { setCellState } from "@/redux/actions";
import { makeConvertSpeedToNumber, makeGetAlive } from "@/redux/selectors";

export interface CellProps {
  isAlive: boolean;
  x: number;
  y: number;
  cellSize: number;
  animationSpeed: number;
}

const makeMapStateToProps = () => {
  const convertSpeedToNumber = makeConvertSpeedToNumber();
  const getAlive = makeGetAlive();
  const mapStateToProps = (state: State, props: any) => {
    return {
      cellSize: state.cellSize,
      isAlive: getAlive(state, props),
      animationSpeed: convertSpeedToNumber(state),
    };
  };
  return mapStateToProps;
};

export const CellComponent: FC<CellProps & { dispatch: AppDispatch }> = (
  props
) => {
  const onClick = useCallback(() => {
    props.dispatch(setCellState({ x: props.x, y: props.y }));
  }, []);
  return (
    <CellItem
      className={`cell cell_${props.y}_${props.x}`}
      isAlive={props.isAlive}
      x={props.x}
      y={props.y}
      cellSize={props.cellSize}
      onClick={onClick}
      animationSpeed={props.animationSpeed}
    />
  );
};

export const Cell = connect(makeMapStateToProps)(CellComponent);
