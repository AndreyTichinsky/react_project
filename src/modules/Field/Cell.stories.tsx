import React from "react";
import { Provider } from "react-redux";
import { withKnobs, number, boolean } from "@storybook/addon-knobs";

import { store } from "@/redux";
import { Cell } from "./Cell";

export default {
  title: "Cell",
  decorators: [withKnobs],
};

export const emptyCell = () => {
  return (
    <Provider store={store}>
      <Cell
        isAlive={boolean("isAlive", false)}
        x={number("x", 0)}
        y={number("y", 0)}
        animationSpeed={number("animationSpeed", 500)}
      />
    </Provider>
  );
};
