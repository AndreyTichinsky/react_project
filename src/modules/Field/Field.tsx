import React, { FC } from "react";
import { FieldWrapper } from "./Field.styled";
import { connect } from "react-redux";
import { State } from "@/redux/store";

import { Cell } from "./Cell";

export interface FieldProps {
  field: boolean[][];
  cellSize: number;
}

const mapStateToProps = (state: State) => ({
  field: state.game.fieldState,
  cellSize: state.game.cellSize,
});

export const FieldComponent: FC<FieldProps> = ({ field, cellSize }) => (
  <FieldWrapper cellSize={cellSize} field={field}>
    {field.map((row, y) => [
      ...row.map((_, x) => <Cell key={`${y}_${x}`} x={x} y={y} />),
    ])}
  </FieldWrapper>
);

export const Field = connect(mapStateToProps)(FieldComponent);
