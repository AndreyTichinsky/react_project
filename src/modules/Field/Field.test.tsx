import React from "react";
import renderer from "react-test-renderer";
import { FieldComponent } from "./Field";
import { store } from "@/redux/store";
import { Provider } from "react-redux";

describe("Field", () => {
  it("empty Field test", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <FieldComponent
              field={[
                [false, false],
                [false, false],
              ]}
              cellSize={50}
            />
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
  it("partial filled cell test", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <FieldComponent
              field={[
                [true, false],
                [false, true],
              ]}
              cellSize={50}
            />
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
