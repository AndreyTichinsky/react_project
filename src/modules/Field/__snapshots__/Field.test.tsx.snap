// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field empty Field test 1`] = `
.emotion-4 {
  position: relative;
  width: 100px;
  height: 100px;
  border: 1px solid #000;
}

.emotion-0 {
  width: 10px;
  height: 10px;
  top: 0px;
  left: 0px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

.emotion-1 {
  width: 10px;
  height: 10px;
  top: 0px;
  left: 10px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

.emotion-2 {
  width: 10px;
  height: 10px;
  top: 10px;
  left: 0px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

.emotion-3 {
  width: 10px;
  height: 10px;
  top: 10px;
  left: 10px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

<div
  className="emotion-4"
>
  <div
    className="cell cell_0_0 emotion-0"
    onClick={[Function]}
    x={0}
    y={0}
  />
  <div
    className="cell cell_0_1 emotion-1"
    onClick={[Function]}
    x={1}
    y={0}
  />
  <div
    className="cell cell_1_0 emotion-2"
    onClick={[Function]}
    x={0}
    y={1}
  />
  <div
    className="cell cell_1_1 emotion-3"
    onClick={[Function]}
    x={1}
    y={1}
  />
</div>
`;

exports[`Field partial filled cell test 1`] = `
.emotion-4 {
  position: relative;
  width: 100px;
  height: 100px;
  border: 1px solid #000;
}

.emotion-0 {
  width: 10px;
  height: 10px;
  top: 0px;
  left: 0px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

.emotion-1 {
  width: 10px;
  height: 10px;
  top: 0px;
  left: 10px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

.emotion-2 {
  width: 10px;
  height: 10px;
  top: 10px;
  left: 0px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

.emotion-3 {
  width: 10px;
  height: 10px;
  top: 10px;
  left: 10px;
  position: absolute;
  border: 1px solid #000;
  box-sizing: border-box;
  -webkit-transition: background-color 500ms ease;
  transition: background-color 500ms ease;
  background-color: #ffffff;
}

<div
  className="emotion-4"
>
  <div
    className="cell cell_0_0 emotion-0"
    onClick={[Function]}
    x={0}
    y={0}
  />
  <div
    className="cell cell_0_1 emotion-1"
    onClick={[Function]}
    x={1}
    y={0}
  />
  <div
    className="cell cell_1_0 emotion-2"
    onClick={[Function]}
    x={0}
    y={1}
  />
  <div
    className="cell cell_1_1 emotion-3"
    onClick={[Function]}
    x={1}
    y={1}
  />
</div>
`;
